File: cpp/problem35.cpp
#include <mystd.h>
using namespace std;

class Solution {
public:
    int searchInsert(vector<int>& nums, int target) {
		int b = 0, e = nums.size(), c;
		while( b != e ){
			c = b + (e-b)/2;
			if( target == nums[c] ){ b = c; break; }
			else if( target < nums[c] ) e = c;
			else b = c+1;
		}
		return b;
    }
};


int main(int argc, char const *argv[]){
	Solution sol;
	vector<int> nums;

	nums.assign( {1,3,5,6} );
	cout << sol.searchInsert( nums,5 ) << endl;

	cout << sol.searchInsert( nums,2 ) << endl;

	nums.assign( {1,3,5,6} );
	cout << sol.searchInsert( nums,7 ) << endl;

	nums.assign( {1,3,5,6} );
	cout << sol.searchInsert( nums,0 ) << endl;
	return 0;
}

File: python/problem35.py
class Solution:
    def searchInsert(self, nums, target):
        """
        :type nums: List[int]
        :type target: int
        :rtype: int
        """
        if target <= nums[0]:
                return 0
        if target == nums[-1]:
            return len(nums)-1
        if target > nums[-1]:
                return len(nums)
        if len(nums)<3:
            if target > nums[0] and target < nums[-1]:
                return 1
        
        l = 0
        r = len(nums)-1
        while l<r:
            mid = (l+r)//2
            
            if nums[mid] == target:
                return mid
            if nums[mid] < target:
                if nums[mid+1] > target:
                    return mid+1
                l =mid
            if nums[mid] > target:
                if nums[mid-1] < target:
                    return mid
                r = mid
    